proposals/0642-JavaScriptCore-safe-bundle-URLs.md

```diff
@@ -0,0 +1,64 @@
+---
+title: JavaScriptCore-safe bundle URLs
+authors:
+- Rob Hogan (@robhogan)
+- Moti Zilberman (@motiz88)
+date: 2023-05-02
+---
+
+# RFC0642: JavaScriptCore-safe bundle URLs
+
+## Summary
+
+Add support for a query-string-free URL format to bundlers, so that symbolication (etc.) is not broken by JavaScriptCore error stack manipulation introduced in https://github.com/WebKit/WebKit/commit/71985aa1520560154a329e5b54159fff3173cf22 (first appearing in iOS 16.4).
+
+## Basic example
+
+`http://localhost:8080/project/index.bundle;&platform=ios`
+
+Should be supported by Metro (and other bundlers) equivalently and in addition to the current query-string form:
+
+`http://localhost:8080/project/index.bundle?platform=ios`
+
+## Motivation
+
+Recent versions of JavaScriptCore strip query strings from source URLs in stack traces (`Error.prototype.stack`).
+
+To implement features such as Metro's `/symbolicate` endpoint, we must be able to derive (a close approximation of) the source that was executed the client. To do so correctly we need to know the build parameters the client used originally. Those parameters being stripped (and defaults used instead) has manifested as obscure resolution errors for React Native users, on both Expo and the community CLI ([1](https://github.com/facebook/react-native/issues/36794), [2](https://github.com/expo/expo/issues/22008)).
+
+Ideally, for maximal compatibility with debuggers (etc.) that may intercept the raw error stack, the source URLs in the error stack should be sufficient, and should be valid URLs, without any post-processing or external context.
+
+## Detailed design
+
+Use a character sequence from the "reserved" set specified in RFC 3986, namely `;&`, as an alternative delimiter in place of `?`, and percent-encode any `?` characters appearing within the subsequent query string, so that the result is a valid [`path`](https://www.rfc-editor.org/rfc/rfc3986#section-3.3).
+
+If the client sends a URL in this format, Metro should respond as if `?` was seen in place of the alternative delimiter, with the following exceptions:
```

Comment:
@motiz88 I think we need to specify what "in this format" means here. For example, what format is this?
```
`/path/with-;&-unencoded/index.bundle?platform=ios`
```

If we interpret it as a "URL of mixed format" as suggested [below](https://github.com/react-native-community/discussions-and-proposals/pull/646/files#diff-0c487972bcf6173d648df22562629f0284700df0e5351731641cbf5ef18c41b6R37), then we'll break the specific edge case we claim to support, because clients are not presently required to percent-encode this sequence within a bundle request path (and in fact, don't). In practice, we *could* ensure that React Native escapes that sequence in future, but it'd still technically be a semver-breaking change to Metro, albeit in a narrow edge case.

It might be more practical to consider an input to be "in this format" if it contains *both* a `;&` within the path, *and no query component*.

Another possibility is to use a character sequence that can't appear in (normalised) POSIX paths - there aren't many, but maybe `//`?